### DISCUSSION TITLE
kbgrep -- knowledgebase search tool

### DISCUSSION THREAD

**Discussion by pcrockett** [bashly-framework/bashly#515] [Show and tell]
https://github.com/pcrockett/kbgrep

some more interesting details:

* i use a fairly minimalist [bats](https://github.com/bats-core/bats-core) test suite, with great test coverage
* i don't actually maintain a `bashly.yml` file -- i keep my config in a [cuelang](https://cuelang.org/) file, and convert it to yaml just before calling `bashly generate`. _because i can 😎 ._

on the to-do list:

* refactoring -- most of my logic is in one file, with a fair amount of code duplication
* refinement -- after using it for some time i'm sure i'll make it even more ergonomic

feedback welcome

**Comment by DannyBen** [comment]
Nice. So it is a wrapper around ripgrep?
What are the advantages over using ripgrep or the silver searcher directly?

And what are the advantages of using cue instead of YAML directly? I am suspicious of people choosing JSON over YAML... :)

When running it on my knowledge base, the output was unexpected.

1. It gave me just the filenames that match - and I expected to see the snippet/line in the file with the matching, sort of like a google search.
2. It seems like the script does not handle files/directories with spaces in them. It gave me a lot of errors like this:
   ```
   rg: Development/NodeJS/Build: No such file or directory (os error 2)
   rg: Global: No such file or directory (os error 2)
   rg: Package.md: No such file or directory (os error 2)

   ```
   when I have a file named `Development/NodeJS/Build Global Package.md`


Also - since I assume this is a markdown-oriented search tool, perhaps employ a markdown display tool, such as [glow](https://github.com/charmbracelet/glow) or [tty-merkdown-cli](https://github.com/piotrmurach/tty-markdown-cli) instead of bat?

Perhaps the tool to use can be defined with an environment variable?

**Reply by pcrockett** [reply]
what kind of maniac puts _spaces_ in file names?? 😁 thanks for spotting that one, I fixed it...

**displaying results:**

you might be interested in the `--select` option that uses `fzf` and `bat`. ctrl+j and ctrl+k will scroll the preview window.

i do think i'll create an env variable to customize the display of results. that said, the way i'm using ripgrep, it's tricky to display results line-by-line... but usually for a knowledgebase i'm just trying to find a file, not a specific snippet.

cool idea using `glow` -- i think i'll leave `bat` as the default because it can do any file type, not just markdown. the preview env variable thing that i haven't built yet should handle that.

**advantages over the greppy tools:**

>  What are the advantages over using ripgrep or the silver searcher directly?

i tried to clarify the rationale in the readme, but i think perhaps i need to use an example:

for the following markdown file:

```markdown
# Why Not YAML 2023-01-12

<!-- [tag:why-not-yaml-2023-01-12] -->

<https://ruudvanasseldonk.com/2023/01/11/the-yaml-document-from-hell>

Next time someone asks me why I hate YAML.

## Links

* \[ref:nickel-2023-05-18]

<!--
    [ref:kb.dev]
    [ref:article]
-->
```

...i can find that file easily with this command:

```bash
kbg kb.dev article yaml
```

if i wanted to use normal grep-like tools, i'd have to remember more details about that specific note (such as the title of the article).

> And what are the advantages of using cue instead of YAML directly?

see [above article](https://ruudvanasseldonk.com/2023/01/11/the-yaml-document-from-hell) 😁 

this was mostly just a fun experiment with cuelang. on a previous bashly project [i used a yaml config file](https://github.com/pcrockett/gramps/blob/main/src/bashly.yml) without issue, and it would have been fine for this project too. but to be honest i find YAML merely _tolerable_ -- i don't enjoy it.

**Reply by DannyBen** [reply]
> what kind of maniac puts spaces in file names?? 😁 thanks for spotting that one, I fixed it...

Heh. My knowledge base is usually displayed with [madness](https://madness.dannyb.co/), so the directory name is the same as the human readable title. I am anti-aliasing wherever possible.

> you might be interested in the --select

Yes, this was slightly more aligned with my expectations.

> see above article 😁

I didn't read all the article, but it seems to take edge cases and use them to "prove" that YAML is not human readable. Almost all the cases he demonstrates in his version of "YAML from hell" can be solved by putting the values in quotes.... like he is already forced to do in JSON....

If you look at the simple, base comparison - YAML is like JSON, without unnecessary punctuations and with more optional functionality.

**Reply by pcrockett** [reply]
Yeah, I probably give YAML a harder time than it deserves... ultimately it comes down to preference, which is why I don't make much of a fuss about it.

**Comment by DannyBen** [comment]
Hey @pcrockett - I just noticed that kbgrep is not linked from https://bashly.dannyb.co/examples/#real-world-examples
Can I link?

**Reply by pcrockett** [reply]
Yup. For what it's worth, it has changed a lot since we last talked about it.
